apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1.IntegrationTests/Emulator/metadata_server.py

```diff
@@ -0,0 +1,353 @@
+    components = indexed_list.split('/')
+    def process(data, i):
+      component = components[i]
+      if component == '*':
+        for item in data:
+          process(item, i + 1)
+      elif i == len(components) - 1:
+        data[component] = IndexedList(data[component])
+      else:
+        process(data[component], i + 1)
+
+    process(all_data, 0)
+
+not_found_error = formatResponse('Error: Not found', 404, common.content_type_html)
+bad_request_error = formatResponse('Error: bad request', 400, common.content_type_html)
+
+@app.after_request
+def add_headers(response):
+  response.headers[common.server] = common.server_value
+
+  # When the token is requested, the response should not have the Metadata-Flavor header
+  if not response.data.startswith('{"access_token":"') and not response.data.startswith('access_token '):
+    response.headers[common.metadata_flavor] = common.metadata_flavor_google
+  return response
+
+@app.route('/', methods = ['GET'])
+def root():
+  return formatResponse('0.1/\ncomputeMetadata/\n', 200, common.content_type_text)
+
+@app.route('/computeMetadata/', methods = ['GET'])
+def computeMetadataRoot():
+  return formatResponse('v1/\nv1beta1/\n', 200, common.content_type_text)
+
+@app.route('/computeMetadata/v1/', methods = ['GET'])
+def getRootData():
+  return getData('')
+
+@app.route('/computeMetadata/v1/<path:path>', methods = ['GET'])
+def getData(path):
+  # TODO There are 500 max hanging gets (not sure if we need to enforce this)
+  # TODO Actually handle waiting for changes and last_etag
+
+  def asBool(value):
+    return value.lower() in ['true', 'on', 'yes', '1']
+
+  def asInt(value, default):
+    try:
+      if value:
+        return int(value)
+      else:
+        return default
+    except ValueError:
+      return default
+
+  def convertToTextRecursive(prefix, o):
+    text = ''
+    if isinstance(o, list):
+      for val in o:
+        text += convertToTextRecursive(prefix, val)
+    elif isinstance(o, IndexedList):
+      for i in range(len(o)):
+        text += convertToTextRecursive(prefix + ('/' if len(prefix) != 0 else '') + unicode(i),
+                                       o[i])
+    elif isinstance(o, OrderedDict):
+      for key in o:
+        if excludeKeyInRecursiveGet(key):
+          continue
+        text += convertToTextRecursive(prefix + ('/' if len(prefix) != 0 else '') + key,
+                                       o[key])
+    elif prefix:
+      text += prefix + ' ' + unicode(o) + '\n'
+    else:
+      text += unicode(o)
+
+    return text
+
+  alt = request.args.get('alt', '')
+  recursive = asBool(request.args.get('recursive', ''))
+  timeout_sec = asInt(request.args.get('timeout_sec', ''), -1)
+  wait_for_change = asBool(request.args.get('wait_for_change', ''))
+
+  components = path.split('/')
+  result = all_data
+  ends_with_slash = False
+  is_token_request = False
+
+  for component in components:
+    if len(component) == 0:
+      ends_with_slash = True
```

Comment:
Yes, that is why ends_with_slash is reset to False if we don't get it here. Something like /a//b/c is valid (I've added another test for this), so the flag gets temporarily set to True, then reset when to gets to 'b'. It will only be True if the last component is empty. But I agree that it should be changed because its not semantically accurate for that temporary period.